File: backend/golang/internal/multiapp/platform_bootstrap.go
package multiapp

import (
	"context"
	"fmt"
	"log"

	"golang.org/x/crypto/bcrypt"

	"rocket-backend/internal/store"
)

// PlatformBootstrap creates the management tables (_apps, _platform_users, _platform_refresh_tokens)
// and seeds a default platform admin user.
func PlatformBootstrap(ctx context.Context, s *store.Store) error {
	if _, err := s.DB.ExecContext(ctx, s.Dialect.PlatformTablesSQL()); err != nil {
		return fmt.Errorf("bootstrap platform tables: %w", err)
	}
	if err := migratePlatformTables(ctx, s); err != nil {
		return fmt.Errorf("migrate platform tables: %w", err)
	}
	if err := seedPlatformAdmin(ctx, s); err != nil {
		return fmt.Errorf("seed platform admin: %w", err)
	}
	return nil
}

// migratePlatformTables adds columns that were introduced after the initial schema.
func migratePlatformTables(ctx context.Context, s *store.Store) error {
	cols, err := s.Dialect.GetColumns(ctx, s.DB, "_apps")
	if err != nil {
		return nil // table may not exist yet, CREATE TABLE IF NOT EXISTS handles it
	}
	if _, ok := cols["db_driver"]; !ok {
		defaultDriver := s.Dialect.Name()
		_, err := s.DB.ExecContext(ctx,
			fmt.Sprintf("ALTER TABLE _apps ADD COLUMN db_driver TEXT NOT NULL DEFAULT '%s'", defaultDriver))
		if err != nil {
			return fmt.Errorf("add db_driver column: %w", err)
		}
		log.Printf("Migrated _apps: added db_driver column (default: %s)", defaultDriver)
	}
	return nil
}

func seedPlatformAdmin(ctx context.Context, s *store.Store) error {
	var count int
	err := s.DB.QueryRowContext(ctx, "SELECT COUNT(*) FROM _platform_users").Scan(&count)
	if err != nil {
		return err
	}
	if count > 0 {
		return nil
	}

	hashBytes, err := bcrypt.GenerateFromPassword([]byte("changeme"), bcrypt.DefaultCost)
	if err != nil {
		return err
	}
	hash := string(hashBytes)

	pb := s.Dialect.NewParamBuilder()
	rolesParam := s.Dialect.ArrayParam([]string{"platform_admin"})
	if s.Dialect.UUIDDefault() == "" {
		// SQLite: generate UUID in Go since there's no gen_random_uuid()
		id := store.GenerateUUID()
		_, err = s.DB.ExecContext(ctx,
			fmt.Sprintf("INSERT INTO _platform_users (id, email, password_hash, roles) VALUES (%s, %s, %s, %s)",
				pb.Add(id), pb.Add("platform@localhost"), pb.Add(hash), pb.Add(rolesParam)),
			pb.Params()...)
	} else {
		_, err = s.DB.ExecContext(ctx,
			fmt.Sprintf("INSERT INTO _platform_users (email, password_hash, roles) VALUES (%s, %s, %s)",
				pb.Add("platform@localhost"), pb.Add(hash), pb.Add(rolesParam)),
			pb.Params()...)
	}
	if err != nil {
		return err
	}

	log.Println("WARNING: Default platform admin created (platform@localhost / changeme) — change the password immediately.")
	return nil
}

File: backend/elixir-phoenix/lib/rocket/multiapp/platform_bootstrap.ex
defmodule Rocket.MultiApp.PlatformBootstrap do
  @moduledoc "Bootstrap platform tables in the management database."

  alias Rocket.Store
  require Logger

  def bootstrap(conn) do
    dialect = Store.dialect()

    dialect.platform_tables_sql()
    |> String.split(";")
    |> Enum.map(&String.trim/1)
    |> Enum.reject(&(&1 == ""))
    |> Enum.each(fn sql ->
      case Store.exec(conn, sql, []) do
        {:ok, _} -> :ok
        {:error, err} -> Logger.error("Platform bootstrap SQL error: #{inspect(err)}")
      end
    end)

    migrate_apps_table(conn, dialect)
    seed_platform_admin(conn, dialect)
  end

  defp migrate_apps_table(conn, dialect) do
    # Add db_driver column if missing (migration for existing installations)
    cols = dialect.get_columns(conn, "_apps")

    if !Map.has_key?(cols, "db_driver") do
      default_driver = dialect.name()

      case Store.exec(conn, "ALTER TABLE _apps ADD COLUMN db_driver TEXT NOT NULL DEFAULT '#{default_driver}'", []) do
        {:ok, _} ->
          Logger.info("Migrated _apps: added db_driver column (default: #{default_driver})")

        {:error, err} ->
          Logger.error("Failed to add db_driver column: #{inspect(err)}")
      end
    end
  end

  defp seed_platform_admin(conn, dialect) do
    case Store.query_row(conn, "SELECT COUNT(*) as count FROM _platform_users", []) do
      {:ok, %{"count" => count}} when is_integer(count) and count > 0 ->
        :ok

      _ ->
        hash = Bcrypt.hash_pwd_salt("changeme")
        roles = dialect.array_param(["platform_admin"])

        {sql, params} =
          if dialect.uuid_default() == "" do
            # SQLite: no gen_random_uuid(), generate UUID manually
            id = Ecto.UUID.generate()

            {"INSERT INTO _platform_users (id, email, password_hash, roles) VALUES ($1, $2, $3, $4)",
             [id, "platform@localhost", hash, roles]}
          else
            {"INSERT INTO _platform_users (email, password_hash, roles) VALUES ($1, $2, $3)",
             ["platform@localhost", hash, roles]}
          end

        case Store.exec(conn, sql, params) do
          {:ok, _} ->
            Logger.warning("Default platform admin created (platform@localhost / changeme) — change the password immediately")

          {:error, err} ->
            Logger.error("Failed to seed platform admin: #{inspect(err)}")
        end
    end
  end
end
